File: java/src/main/java/com/github/digitalheir/simple/UnicodeCharacter.java
package com.github.digitalheir.simple;

import com.github.digitalheir.unicode.Bmp;
import com.github.digitalheir.unicode.Character;
import com.github.digitalheir.unicode.Description;
import com.github.digitalheir.unicode.Elsevier;
import com.github.digitalheir.unicode.Entity;
import com.github.digitalheir.unicode.Surrogate;
import com.github.digitalheir.unicode.Wolfram;

import java.util.List;
import java.util.stream.Stream;

import static java.util.Objects.nonNull;
import static java.util.Objects.requireNonNull;
import static java.util.stream.Collectors.toList;

public class UnicodeCharacter {
  // [number] | [number, number] | [number, number, number]
  public final List<Integer> dec;

  public final String description;
  public final String descriptionUnicodeVersion;
  public final String mode;
  public final String type;
  public final String image;
  public final String afii;
  public final String latex;
  public final String varlatex;
  public final String mathlatex;
  public final String ams;
  public final String aps;
  public final String acs;
  public final String aip;
  public final String ieee;
  public final String springer;
  public final String comment;

  public final String wolfram;
  public final String wolframId;

  public final SimpleSurrogate surrogate;
  public final String bmp;

  public final String elsevierGrid;
  public final String elsevierElsrender;
  public final String elsevierEnt;
  public final String elsevierDesc;

  public final List<SimpleEntity> entity;

  public final List<SimpleFont> font;


  @SuppressWarnings("UnnecessaryThis")
  public UnicodeCharacter(
      List<Integer> dec,
      String mode,
      String type,
      String image,
      String afii,
      String latex,
      String varlatex,
      String mathlatex,
      Elsevier elsevier,
      String ams,
      String aps,
      String acs,
      String aip,
      String ieee,
      Wolfram wolfram,
      String springer,
      List<Entity> entity,
      List<SimpleFont> font,
      String comment,
      Surrogate surrogate,
      Bmp bmp,
      Description description) {
    this.dec = requireNonNull(dec);
    this.mode = mode;
    this.type = type;
    this.image = image;
    this.afii = (afii);
    this.latex = (latex);
    this.varlatex = (varlatex);
    this.mathlatex = (mathlatex);

    this.elsevierGrid = (elsevier == null ? null : elsevier.getGrid());
    this.elsevierElsrender = (elsevier == null ? null : elsevier.getElsrender());
    this.elsevierEnt = (elsevier == null ? null : elsevier.getEnt());
    this.elsevierDesc = (elsevier == null ? null : elsevier.getDesc());

    this.ams = (ams);
    this.aps = (aps);
    this.acs = (acs);
    this.aip = (aip);
    this.ieee = (ieee);

    this.wolfram = nonNull(wolfram) ? wolfram.getvalue() : null;
    this.wolframId = nonNull(wolfram) ? wolfram.getId() : null;

    this.springer = (springer);
    this.entity = entity != null ? entity.stream().map(SimpleEntity::new).collect(toList()) : null;
    this.font = font;
    this.comment = (comment);
    this.surrogate = surrogate != null ? new SimpleSurrogate(surrogate) : null;
    this.bmp = bmp != null ? ((Character) bmp.getRef()).getId() : null;

    // requireNonNull(description);
    this.description = description.getvalue();
    this.descriptionUnicodeVersion = description.getUnicode();
  }

  public UnicodeCharacter(com.github.digitalheir.unicode.Character c) {
    this(

        Stream.of(c.dec.split("-"))
            .map(Integer::new)
            .collect(toList()),

        c.mode,
        c.type,
        c.image,
        c.afii,
        c.latex,
        c.varlatex,
        c.mathlatex,
        c.elsevier,
        c.ams,
        c.aps,
        c.acs,
        c.aip,
        c.ieee,
        c.wolfram,
        c.springer,
        c.entity,
        c.font == null ? null : c.font.stream().map(SimpleFont::new).collect(toList()),
        c.comment,
        c.surrogate,
        c.bmp,
        c.description
    );
  }
}

File: typescript/src/UnicodeCharacter.ts
import {decimal2unicode} from "./characters/decimal2unicode";
import {UnicodeIdentification} from "./characters/unicodeIdentification";

export interface Surrogate {
    mathvariant: string;
    ref: string;
}

export interface Entity {
    id: string;
    set: string;
    desc?: string;
    multipleShapesExcluded?: boolean;
    xref?: string;
}

export interface Font {
    name: string;
    pos: number | "40)";
}

export const startingU = /^U/;

export function determineHex(u: UnicodeCharacter, withPrefix: boolean = true): string {
    return u._id.replace(startingU, withPrefix ? "0x" : "");
}

export function prettyPrintCodePoint(u: UnicodeCharacter): string {
    return u._id.replace(startingU, "U+");
}

export interface UnicodeCharacter {
    _id: string;
    description: string;
    dec: [number] | [number, number] | [number, number, number];

    descriptionUnicodeVersion?: string;
    mode?: string ;
    type?: string ;
    image?: string ;
    afii?: string ;
    latex?: string ;
    varlatex?: string ;
    mathlatex?: string ;
    ams?: string ;
    aps?: string ;
    acs?: string ;
    aip?: string ;
    ieee?: string ;
    springer?: string ;
    wolfram?: string ;
    wolframId?: string ;
    elsevierGrid?: string ;
    elsevierElsrender?: string ;
    elsevierEnt?: string ;
    elsevierDesc?: string ;
    comment?: string;
    bmp?: string ;

    entity?: Entity[] ;
    font?: Font[] ;
    surrogate?: Surrogate ;

}


export function normalizeStrings(c: UnicodeCharacter): string[]{
    const base: string[] = [];

    base.push(
        "id",
        "unicode",
        "codepoint",
        c._id,
        determineHex(c, true),
        determineHex(c, false),
        prettyPrintCodePoint(c),
    );

    if(!!c.dec) {
        base.push("decimal", "dec");
        base.push(...c.dec.map(c=>c.toString()));
    }

    if(!!c.description) base.push("description", c.description);
    if(!!c.descriptionUnicodeVersion) base.push("version", c.descriptionUnicodeVersion);
    if(!!c.mode) base.push("mode", c.mode);
    if(!!c.type) base.push("type", c.type);
    if(!!c.image) base.push("image", c.image);
    if(!!c.afii) base.push("afii", c.afii);
    if(!!c.latex) base.push("latex", c.latex);
    if(!!c.varlatex) base.push("varlatex", c.varlatex);
    if(!!c.mathlatex) base.push("mathlatex", c.mathlatex);
    if(!!c.ams) base.push("ams", c.ams);
    if(!!c.aps) base.push("aps", c.aps);
    if(!!c.acs) base.push("acs", c.acs);
    if(!!c.aip) base.push("aip", c.aip);
    if(!!c.ieee) base.push("ieee", c.ieee);
    if(!!c.springer) base.push("springer", c.springer);
    if(!!c.wolfram) base.push("wolfram", c.wolfram);
    if(!!c.wolframId) base.push("wolfram", c.wolframId);
    if(!!c.elsevierGrid) base.push("elsevier","grid", c.elsevierGrid);
    if(!!c.elsevierElsrender) base.push("elsevier","elsrender", c.elsevierElsrender);
    if(!!c.elsevierEnt) base.push("elsevier","ent", c.elsevierEnt);
    if(!!c.elsevierDesc) base.push("elsevier", "description", c.elsevierDesc);
    if(!!c.comment) base.push("comment", c.comment);
    if(!!c.bmp) base.push("bmp", c.bmp);

    if(!!c.entity){
        base.push("entity");
        base.push(...c.entity.map(e => e.id));
        // todo more string for entity
    }

    if(!!c.font){
        base.push("font");
        base.push(...c.font.map(e => e.name));
        base.push(...c.font.map(e => e.pos.toString()));
    }

    if(!!c.surrogate){
        base.push("surrogate");
        if(c.surrogate.mathvariant) base.push(c.surrogate.mathvariant);
        // todo also index ref?
    }

    return base.map(s => s.toLowerCase());
}


export function getUnicodeIdentifierFromDecimal(...codepoints: number[]): UnicodeIdentification | undefined {
    return <UnicodeIdentification>decimal2unicode[codepoints.join("-")];
}
